src/Microsoft.SqlTools.ServiceLayer/FileBrowser/FileBrowserService.cs

```diff
@@ -0,0 +1,222 @@
+using Microsoft.SqlTools.ServiceLayer.Hosting;
+using Microsoft.SqlTools.ServiceLayer.FileBrowser.Contracts;
+using Microsoft.SqlTools.ServiceLayer.FileBrowser.FileValidator;
+using System.Collections.Generic;
+
+namespace Microsoft.SqlTools.ServiceLayer.FileBrowser
+{
+    /// <summary>
+    /// Main class for file browser service
+    /// </summary>
+    public sealed class FileBrowserService
+    {
+        private static readonly Lazy<FileBrowserService> LazyInstance = new Lazy<FileBrowserService>(() => new FileBrowserService());
+        public static FileBrowserService Instance => LazyInstance.Value;
+
+        // Caches file browser operation for expanding node request
+        private Dictionary<string, FileBrowserOperation> ownerToFileBrowserMap = new Dictionary<string, FileBrowserOperation>();
+        private ConnectionService connectionService = null;
+
+        internal ConnectionService ConnectionServiceInstance
+        {
+            get
+            {
+                if (connectionService == null)
+                {
+                    connectionService = ConnectionService.Instance;
+                }
+                return connectionService;
+            }
+            set
+            {
+                connectionService = value;
+            }
+        }
+
+        /// <summary>
+        /// Initializes the service instance
+        /// </summary>
+        /// <param name="serviceHost"></param>
+        /// <param name="context"></param>
+        public void InitializeService(ServiceHost serviceHost)
+        {
+            // Open a file browser
+            serviceHost.SetRequestHandler(FileBrowserOpenRequest.Type, HandleFileBrowserOpenRequest);
+
+            // Expand a folder node
+            serviceHost.SetRequestHandler(FileBrowserExpandRequest.Type, HandleFileBrowserExpandRequest);
+
+            // Validate the selected files
+            serviceHost.SetRequestHandler(FileBrowserValidateRequest.Type, HandleFileBrowserValidateRequest);
+
+            // Close the file browser
+            serviceHost.SetRequestHandler(FileBrowserCloseRequest.Type, HandleFileBrowserCloseRequest);
+        }
+
+        internal async Task HandleFileBrowserOpenRequest(
+            FileBrowserOpenParams fileBrowserParams,
+            RequestContext<FileBrowserOpenResponse> requestContext)
+        {
+            FileBrowserOpenResponse response = new FileBrowserOpenResponse();
+
+            try
+            {
+                ConnectionInfo connInfo;
+                this.ConnectionServiceInstance.TryFindConnection(
+                    fileBrowserParams.OwnerUri,
+                    out connInfo);
+
+                if (connInfo != null)
+                {
+                    SqlConnection sqlConn = GetSqlConnection(connInfo);
+                    FileBrowserOperation browser = new FileBrowserOperation(sqlConn, fileBrowserParams.ExpandPath, fileBrowserParams.FileFilters);
+                    browser.PopulateFileTree();
```

Comment:
>browser.PopulateFileTree(); [](start = 20, length = 27)

is it fine to run this method synchronously in an method handler?  i.e., can we guarantee it will always complete in less than 500ms?

Reply:
changed to async.